pkg/requests/builder.go

```
@@ -0,0 +1,173 @@
+	}
+
+	req, err := http.NewRequestWithContext(r.context, r.method, r.endpoint, r.body)
+	if err != nil {
+		return nil, fmt.Errorf("error creating request: %v", err)
+	}
+	req.Header = r.header
+
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return nil, fmt.Errorf("error performing request: %v", err)
+	}
+
+	r.response = resp
+	return resp, nil
+}
+
+// Into performs the request and attempts to unmarshal the response into the
+// the given interface. The response body is assumed to be JSON.
+// The response must have a 200 status otherwise an error will be returned.
+func (r *builder) Into(into interface{}) error {
+	resp, err := r.Do()
+	if err != nil {
+		return err
+	}
+
+	return Into(resp, into)
+}
+
+// JSON performs the request and attempts to unmarshal the response into a
+// simplejson.Json. The response body is assume to be JSON.
+// The response must have a 200 status otherwise an error will be returned.
+func (r *builder) JSON() (*simplejson.Json, error) {
+	resp, err := r.Do()
+	if err != nil {
+		return nil, err
+	}
+
+	body, err := getResponseBody(resp)
+	if err != nil {
+		return nil, err
+	}
+
+	data, err := simplejson.NewJson(body)
+	if err != nil {
+		return nil, fmt.Errorf("error reading json: %v", err)
+	}
+	return data, nil
+}
+
+// Into attempts to unmarshal the response into the the given interface.
+// The response body is assumed to be JSON.
+// The response must have a 200 status otherwise an error will be returned.
+func Into(resp *http.Response, into interface{}) error {
+	body, err := getResponseBody(resp)
+	if err != nil {
+		return err
+	}
+
+	if err := json.Unmarshal(body, into); err != nil {
+		return fmt.Errorf("error unmarshalling body: %v", err)
+	}
+
+	return nil
+}
+
+// getResponseBody extracts the response body, but will only return the body
+// if the response was successful.
+func getResponseBody(resp *http.Response) ([]byte, error) {
+	body, err := ioutil.ReadAll(resp.Body)
+	resp.Body.Close()
```

Do we want to handle this error? I've had it trained into me to handle all errors (my IDE likes to yell at me about it) -- but I even seen go underlying `http` library ignoring error handling on Body.Close()

---

Had a dig through the code, given we have done an `ioutil.ReadAll` just above, there's no need to check the error as the code path checks whether the body has been read, because it is at EOF, it returns a non-nil error. What we should do is check the error from the line above first, the typical pattern is to defer the response body close so I'll do that